File: sdks/wolkpay-python/wolkpay/resources/webhooks.py
"""
Webhooks Resource
"""

from typing import Optional, Dict, Any, List, TYPE_CHECKING

from ..types import WebhookEvent

if TYPE_CHECKING:
    from ..client import WolkPay


class WebhooksResource:
    """
    Webhooks API Resource
    
    Configure and manage webhook settings.
    """
    
    def __init__(self, client: "WolkPay"):
        self.client = client
    
    def get_events(
        self,
        delivered: Optional[bool] = None,
        event_type: Optional[str] = None,
        page: int = 1,
        limit: int = 20
    ) -> List[WebhookEvent]:
        """
        List webhook events
        
        Args:
            delivered: Filter by delivery status
            event_type: Filter by event type
            page: Page number
            limit: Items per page
            
        Returns:
            List of WebhookEvent objects
            
        Example:
            >>> events = client.webhooks.get_events(delivered=False)
            >>> for event in events:
            ...     print(f"{event.event_type}: {event.payment_id}")
        """
        params = {"page": page, "limit": limit}
        
        if delivered is not None:
            params["delivered"] = delivered
        if event_type:
            params["event_type"] = event_type
        
        response = self.client.get("webhooks/events", params=params)
        
        if isinstance(response, dict) and "data" in response:
            return [WebhookEvent.from_dict(e) for e in response["data"]]
        elif isinstance(response, list):
            return [WebhookEvent.from_dict(e) for e in response]
        
        return []


# Webhook event types
class WebhookEventTypes:
    """Available webhook event types"""
    
    # Payment events
    PAYMENT_CREATED = "payment.created"
    PAYMENT_PENDING = "payment.pending"
    PAYMENT_PROCESSING = "payment.processing"
    PAYMENT_COMPLETED = "payment.completed"
    PAYMENT_FAILED = "payment.failed"
    PAYMENT_EXPIRED = "payment.expired"
    PAYMENT_CANCELLED = "payment.cancelled"
    PAYMENT_REFUNDED = "payment.refunded"
    
    # PIX specific
    PIX_RECEIVED = "pix.received"
    PIX_CONFIRMED = "pix.confirmed"
    
    # Crypto specific
    CRYPTO_DETECTED = "crypto.detected"
    CRYPTO_CONFIRMED = "crypto.confirmed"
    
    @classmethod
    def all(cls) -> List[str]:
        """Get all event types"""
        return [
            cls.PAYMENT_CREATED,
            cls.PAYMENT_PENDING,
            cls.PAYMENT_PROCESSING,
            cls.PAYMENT_COMPLETED,
            cls.PAYMENT_FAILED,
            cls.PAYMENT_EXPIRED,
            cls.PAYMENT_CANCELLED,
            cls.PAYMENT_REFUNDED,
            cls.PIX_RECEIVED,
            cls.PIX_CONFIRMED,
            cls.CRYPTO_DETECTED,
            cls.CRYPTO_CONFIRMED
        ]

File: sdks/wolkpay-php/src/Resources/Webhooks.php
<?php

declare(strict_types=1);

namespace WolkPay\Resources;

use WolkPay\WolkPay;

/**
 * Webhooks Resource
 */
class Webhooks
{
    private WolkPay $client;

    public function __construct(WolkPay $client)
    {
        $this->client = $client;
    }

    /**
     * List webhook events
     *
     * @param array $params Filter parameters
     *   - delivered: bool Filter by delivery status
     *   - event_type: string Filter by event type
     *   - page: int Page number
     *   - limit: int Items per page
     *
     * @return array Array of webhook event objects
     */
    public function listEvents(array $params = []): array
    {
        $queryParams = [
            'page' => $params['page'] ?? 1,
            'limit' => $params['limit'] ?? 20,
        ];

        if (isset($params['delivered'])) {
            $queryParams['delivered'] = $params['delivered'];
        }

        if (isset($params['event_type'])) {
            $queryParams['event_type'] = $params['event_type'];
        }

        $response = $this->client->request('GET', 'webhooks/events', null, $queryParams);

        if (isset($response['data'])) {
            return array_map(fn($e) => (object) $e, $response['data']);
        }

        return array_map(fn($e) => (object) $e, $response);
    }
}

/**
 * Webhook Event Types
 */
class WebhookEventTypes
{
    // Payment events
    public const PAYMENT_CREATED = 'payment.created';
    public const PAYMENT_PENDING = 'payment.pending';
    public const PAYMENT_PROCESSING = 'payment.processing';
    public const PAYMENT_COMPLETED = 'payment.completed';
    public const PAYMENT_FAILED = 'payment.failed';
    public const PAYMENT_EXPIRED = 'payment.expired';
    public const PAYMENT_CANCELLED = 'payment.cancelled';
    public const PAYMENT_REFUNDED = 'payment.refunded';

    // PIX specific
    public const PIX_RECEIVED = 'pix.received';
    public const PIX_CONFIRMED = 'pix.confirmed';

    // Crypto specific
    public const CRYPTO_DETECTED = 'crypto.detected';
    public const CRYPTO_CONFIRMED = 'crypto.confirmed';

    /**
     * Get all event types
     */
    public static function all(): array
    {
        return [
            self::PAYMENT_CREATED,
            self::PAYMENT_PENDING,
            self::PAYMENT_PROCESSING,
            self::PAYMENT_COMPLETED,
            self::PAYMENT_FAILED,
            self::PAYMENT_EXPIRED,
            self::PAYMENT_CANCELLED,
            self::PAYMENT_REFUNDED,
            self::PIX_RECEIVED,
            self::PIX_CONFIRMED,
            self::CRYPTO_DETECTED,
            self::CRYPTO_CONFIRMED,
        ];
    }
}
